Removed Multi-processing and replaced it with `Snakemake's` native multiprocessing support

axiomcura opened the pull request:
This pull request focuses on replacing python's `multiprocessing` module with `snakemake`'s native multiprocessing scheduler.

gwaybio commented on .gitignore:
```diff
@@ -2,3 +2,13 @@
+
+.vscode
+configs.bak
+.ipynb
+view.ipynb
```
```suggestion
*.ipynb
```

This should work :)

gwaybio commented on Snakefile:
```diff
@@ -24,4 +24,4 @@ rule all:
             plate_id=PLATE_IDS,
         ),
-        "results/preprocessing/consensus.csv.gz",
+        "results/preprocessing/median_consensus.csv.gz",
```
quick question - should this always be "median"? What if someone used, for example `modz` consensus? Is there a wildcard for this?

axiomcura replied:
I definitely agree! 

There was no reasoning for the `median_consensus` name. It was more for a place holder since my rotation presentation was the next day (I think).

gwaybio commented on configs/analysis_configs/annotate_configs.yaml:
```diff
@@ -1,13 +1,25 @@
     float_format: null
     cmap_args: {}
+
+      # - Metadata_Well
+# challenges
+# -- compression options:
+  # -- compression time changes for each gzip there github might flag changes
+  # occurred in the data
+  #
+  # future: conpressino_options might not be in the config file. define in
+  # global configs (i.e threads)
```
added by accident? If they are helpful, i suggest adding to an issue.

gwaybio commented on configs/analysis_configs/feature_select_configs.yaml:
```diff
@@ -6,18 +6,25 @@ feature_select_configs:
       - drop_na_columns
-      - correlation_threshold
-      # - blocklist This requires a blocklist file. (meeting with greg)
+      # - correlation_threshold
```
typo?

axiomcura replied:
I commented it out in order for the cell health dataset to work. An error kept riseing when using correlation_threshold 

I'll put that in the issues.

gwaybio commented on configs/analysis_configs/feature_select_configs.yaml:
```diff
@@ -6,18 +6,25 @@ feature_select_configs:
+    noise_removal_perturb_groups: null
+    noise_removal_stdev_cutoff: null
+
+
+
+# there is a path but snakemake is not capturing it.
+# also the blocklist_file var might be wrong
+# - blocklist This requires a blocklist file. (meeting with greg)
```
Remove extra lines

gwaybio commented on configs/analysis_configs/normalize_configs.yaml:
```diff
@@ -1,13 +1,20 @@
-    compression_options: "gzip"
+    meta_features: infer
+    # samples: all
+    samples: "Metadata_pert_name == 'EMPTY'"
+    # method: standardize
+    method: mad_robustize
+    compression_options:
+      method: gzip
+      mtime: 1
     float_format: null
     mad_robustize_epsilon: 1.0e-18
     spherize_center: True
-    spherize_method: "ZCA-cor"
+    spherize_method: ZCA-cor
     spherize_epsilon: 1.0e-6
+
+# Notes
+# -- order of different normalization methods
```
Needs some cleaning

gwaybio commented on configs/analysis_configs/single_cell_configs.yaml:
```diff
@@ -1,19 +1,32 @@
+    fields_of_view: all
+    fields_of_view_feature: Metadata_Site
+    object_feature: Metadata_ObjectNumber
+
+
+# linking cols link (Discord) -> notes are there
+# strata: ["Metadata_Plate", "Metadata_Well"]
```
Cleaning required

gwaybio commented on rules/feature_select.smk:
```diff
@@ -24,7 +24,7 @@ rule create_consensus:
         ),
     output:
-        "results/preprocessing/consensus.csv.gz",
+        "results/preprocessing/median_consensus.csv.gz",
```
Maybe this is where my earlier comment belongs?

gwaybio commented on rules/preprocessing.smk:
```diff
@@ -28,49 +28,60 @@ configfile: "configs/configuration.yaml"
-    threads: config["analysis_configs"]["preprocessing"]["threads"]
-    script:
-        "../scripts/aggregate_cells.py"
+    shell:
```
I thought that `python` was the .smk command we should be using here. Is there an advantage to `shell`?

axiomcura replied:
Hmmmm.... Good point. 

I Learned it from this [tutorial]( http://www.hpc-carpentry.org/hpc-python/12-resources/index.html) under the `Managing CPU's section` 

It seems that the `shell` provides parallelization support. I am not too sure. 

I could be wrong. I can try with the `script` rule attribute to see if it works. If it works with the `script` rule attribute, then there will be no use for the `argparse` imports since `snakemake` automatically inputs the given parameters into the script at runtime.

gwaybio commented on scripts/aggregate_cells.py:
```diff
@@ -94,29 +94,37 @@ def aggregate(
-    config_path = itertools.repeat(str(snakemake.params["aggregate_config"]))
-    inputs = list(
-        zip(sqlfiles, meta_data_dir, barcode_map, cell_count_out, aggregate_profile_out, config_path)
+    # CLI arguments
```
One thing I like doing with argparse (or other CLI) is to abstract it into a different function that you import.

Essentially, you can write a function called `grab_args()`, import it here, and then call it: `args = grab_args()`.

then reference them like you do below (e.g. `sql_file = args.input`).

Does this make sense? It has a huge benefit of only needing to maintain one argument parser, and making each `.py` function much more digestible.

axiomcura replied:
I do not thing using `argparse` will be necessary anymore. It seems that the `script` rule attribute does support multiprocessing. 

https://github.com/WayScience/CytoPipe/pull/11#discussion_r916287016

gwaybio commented on scripts/consensus.py:
```diff
@@ -33,14 +38,171 @@ def concatenate_data(profile_list: list) -> pd.DataFrame:
     concat_df = concat_df.drop(na_cols, axis="columns")
 
-    return concat_df
+    # dropping Costes Features
+    costes_cols_to_drop = [x for x in concat_df.columns if "costes" in x.lower()]
+    concat_df = concat_df.drop(costes_cols_to_drop, axis="columns")
```
We shouldn't hardcode this. (IIRC, costes features are only broken in one iteration of CP)

axiomcura replied:
I remember you have mentioned a function that deals with `costes` without the hard coding. I  completely forgot the name

gwaybio replied:
you can drop all hardcoded examples. They shouldn't be introduced into the `main` branch.

gwaybio commented on scripts/consensus.py:
```diff
@@ -33,14 +38,171 @@ def concatenate_data(profile_list: list) -> pd.DataFrame:
+    costes_cols_to_drop = [x for x in concat_df.columns if "costes" in x.lower()]
+    concat_df = concat_df.drop(costes_cols_to_drop, axis="columns")
+
+    x_groupby_cols = ["Metadata_gene_name", "Metadata_pert_name", "Metadata_cell_line"]
```
This needs to be a customizable argument. Not all of these columns will always be present, and sometimes we want to aggregate different columns.

axiomcura replied:
The whole `consensus.py` contains lots of hard coded parameters. Will submit this as an issue

axiomcura replied:
Issue created: #12

gwaybio replied:
good to open the issue, but I don't think we should merge hardcoded variables into main. In other words, it's best to close #12 in this PR.

gwaybio commented on scripts/consensus.py:
```diff
@@ -33,14 +38,171 @@ def concatenate_data(profile_list: list) -> pd.DataFrame:
+        .groupby(x_groupby_cols)
+        .count()
+        .reset_index()
+        .assign(data_type="cell_painting")
```
i don't think we need this

axiomcura replied:
has been removed in: https://github.com/WayScience/CytoPipe/pull/11/commits/2f55798431b9dcd229f6193810b74a40c8211d7b

gwaybio commented on scripts/consensus.py:
```diff
@@ -33,14 +38,171 @@ def concatenate_data(profile_list: list) -> pd.DataFrame:
+            how="left",
+            on=x_groupby_cols,
+        )
+    )
```
Maybe actually delete this whole definition of `x_metacount_df`? I am not sure if it belongs here.

axiomcura replied:
Has been removed in: https://github.com/WayScience/CytoPipe/pull/11/commits/2f55798431b9dcd229f6193810b74a40c8211d7b

gwaybio commented on scripts/consensus.py:
```diff
@@ -1,9 +1,11 @@
+from distutils.command.build import build
```
I started commenting on this script, but then I realized that you might have included it by mistake.

This looks way too specific to the Cell health data than what we are looking for in a generic tool.

axiomcura replied:
Yeah. it's Vscode auto import...

gwaybio commented on scripts/normalize.py:
```diff
@@ -12,26 +14,38 @@ def normalization(anno_file: str, norm_outfile: str) -> None:
     with open(normalize_config_path, "r") as yaml_contents:
         normalize_config = yaml.safe_load(yaml_contents)["normalize_configs"]["params"]
 
+    # meta_features = [
+    #     "Metadata_Plate",
+    #     "Metadata_Well",
+    #     # "Metadata_Plate_Map_Name",
+    #     "Metadata_Plate",
+    #     "Metadata_Well",
+    #     "Metadata_Object_Count",
+    # ]
```
delete all superfluous lines.

gwaybio reviewed (changes requested):
looking good. I'm not sure where to point to in regards to where exactly you've replaced multiprocessing with native support. 

It also looks like you've introduced some other new things (e.g. argparse, and maybe a file by accident).

I made several suggestions about how to make this code a bit cleaner, and you should remove some superfluous lines before merging.

Let me know when it's ready for another peek!

gwaybio commented on scripts/annotate.py:
```diff
@@ -1,11 +1,22 @@
+from ast import arg
```
I think there are still a couple unused imports (e.g. arg and require)

axiomcura replied:
Ugh... I honestly do not know how these automatically get imported...

gwaybio reviewed (changes requested):
It makes sense to update code introduced in one PR in the same PR rather than merging into main.

In other words, concretely:

- We should not add argparse code if we've decided to move away from argparse
- We should not add hardcoded examples only to remove them later

Typically, opening an issue is sufficient for when an out of scope suggestion comes up. These two examples are entirely within scope because they are the exact nature of the suggested change.

axiomcura commented:
>good to open the issue, but I don't think we should merge hardcoded variables into main. In other words, it's best to close https://github.com/WayScience/CytoPipe/issues/12 in this PR.

Does this call for renaming the PR that acknowledges both `snakemake` native multiprocessing and setting defaults for consensus?

gwaybio commented:
> Does this call for renaming the PR that acknowledges both snakemake native multiprocessing and setting defaults for consensus?

If you feel that appropriate, sounds good to me!

axiomcura commented:
Hey @gwaybio !

I have edited the consensus script along with the configuration file implementation on this commit: https://github.com/WayScience/CytoPipe/pull/11/commits/ccb111e32e8ed3807125ba74e576391ce1dc6116

Let me know if the steps taken within the script is the correct way of producing a consensus profile!

gwaybio commented on Snakefile:
```diff
@@ -4,24 +4,22 @@ import glob
             plate_id=PLATE_IDS,
         ),
-        "results/preprocessing/consensus.csv.gz",
+        "results/preprocessing/median_consensus.tsv.gz",
```
still need to fix this (I think how you had it before was better?)

e.g. what if the user performed `modz` consensus?

axiomcura replied:
Before it was `consensus`, a more general name. I can revert it back. Still figuring out how to switch the output name dynamically, since `snakemake` is very `strict` about the names of output files!

gwaybio commented on configs/analysis_configs/consensus_configs.yaml:
```diff
@@ -0,0 +1,13 @@
+consensus_config:
```
hmm, I'm still a bit confused by these parameters. Where did you take them from?

Shouldn't the parameters for consensus (like the other steps) include only the arguments in [`pycytominer.consensus`](https://github.com/cytomining/pycytominer/blob/master/pycytominer/consensus.py)?

axiomcura replied:
this came from the aggregate function: https://github.com/cytomining/pycytominer/blob/master/pycytominer/aggregate.py#L14

but I will replace it! I didn't know there was a consensus function

gwaybio commented on scripts/consensus.py:
```diff
@@ -33,21 +45,44 @@ def concatenate_data(profile_list: list) -> pd.DataFrame:
+        float_format=consensus_aggregate_config["float_format"],
+    )
+
+    # Output Profile Mapping for Downstream Analysis
+    profile_id_mapping_df = x_median_df.loc[
+        :, x_median_df.columns.str.startswith("Metadata")
+    ]
+    file = os.path.join("data", "profile_id_metadata_mapping.tsv")
+    profile_id_mapping_df.to_csv(file, sep="\t", index=False)
```
I think this is still cell-health specific (line 70-75)

gwaybio commented on scripts/consensus.py:
```diff
@@ -12,11 +16,19 @@ def concatenate_data(profile_list: list) -> pd.DataFrame:
     pd.DataFrame
         concatenated normalized aggregated features
     """
+    # loading config file
+    aggregate_path_obj = Path(config)
```
can you keep the nomenclature consistent?

i.e. `consensus_` instead of `aggregate_`. Using aggregate here is confusing

gwaybio commented on scripts/consensus.py:
```diff
@@ -33,21 +45,44 @@ def concatenate_data(profile_list: list) -> pd.DataFrame:
-    return concat_df
+    # aggregate
+    consensus_aggregate_config = consensus_config["aggregate"]
+    x_median_df = aggregate(
```
ah! I see maybe your confusion. pycytominer has a consensus function:

https://github.com/cytomining/pycytominer/blob/master/pycytominer/consensus.py

Let's use this function instead of aggregate

gwaybio commented on scripts/consensus.py:
```diff
@@ -33,21 +45,44 @@ def concatenate_data(profile_list: list) -> pd.DataFrame:
 
-    return concat_df
+    # aggregate
+    consensus_aggregate_config = consensus_config["aggregate"]
```
why not just `consensus_config` as the variable name?

gwaybio reviewed (changes requested):
So close - i think one more quick iteration and we'll be able to merge. Happy to jump on a call to clarify any points

axiomcura commented:
Hello @gwaybio 

Hopefully this take's care of it! 

Removing cell health specific code: 91c11addac3f9c1e3211dee13c90ea6d28155d1f
Replaced aggregate function with consensus function in the `consensus.py` script: 76a371380b4dc645152e193c16f2f7585f23c2fc
reverted consensus output name to `consensus.tsv`: https://github.com/WayScience/CytoPipe/pull/11/commits/a547f52d4c0c0e093a3b331dab80b321e46a7d74

gwaybio commented on scripts/consensus.py:
```diff
@@ -33,21 +45,36 @@ def concatenate_data(profile_list: list) -> pd.DataFrame:
 
-    return concat_df
+    # generating consensus profile
+    x_median_df = consensus(
```
```suggestion
    x_consensus_df = consensus(
```

No requirement that it is median

gwaybio commented on scripts/consensus.py:
```diff
@@ -33,21 +45,36 @@ def concatenate_data(profile_list: list) -> pd.DataFrame:
+    )
+
+    # Saving consensus profile
+    x_median_df.to_csv(consensus_file_out, sep="\t", index=False)
```
```suggestion
    x_consensus_df.to_csv(consensus_file_out, sep="\t", index=False)
```

gwaybio commented on rules/preprocessing.smk:
```diff
@@ -65,9 +59,9 @@ rule annotate:
+        "results/preprocessing/{plate_id}_augmented.csv.gz",
     output:
-        expand("results/preprocessing/{plate_id}_normalized.csv.gz", plate_id=PLATE_IDS),
+        "results/preprocessing/{plate_id}_normalized.csv.gz",
```
Do you need to add a consensus rule in this .smk file?

axiomcura replied:
The consensus rule is in the `rulesfeature_select.smk` not `preprocessing.smk`. Unless you want the `feature_select` and `consensus` rule in one `.smk` file. 

I guess my question is - Is `feature_select` and `consensus` a preprocessing step?

gwaybio replied:
this is a good question, and the answer isn't immediately obvious (to me at least). I'm not sure there even is one!

If I had to guess, I'd say _none_ of this is preprocessing. It is all _processing_.

I think of preprocessing as steps immediately following data extraction from images (illum correction, quality control, the things Jenna has been focused on). I'm not sure where we draw the line between preprocessing and processing, but I think it comes before aggregation even.

Either way, I think separating rules between `rulesfeature_select.smk` and `preprocessing.smk` is a bit confusing (for the reasons I outlined above about how to think about processing vs. preprocessing).

But, I think that fixing this might be beyond scope of this PR - I opened #12 for us to discuss further, and to document that we still need to address this once we merge this present PR.

gwaybio reviewed (approved):
Looks good! One question and one minor suggestion.

Feel free to merge at will once you answer these two things
